pkg/controller/servicemesh/member/controller_test.go

```diff
@@ -0,0 +1,382 @@
+			memberRollExists: true,
+			deletion:         true,
+			reactor:          test.ClientFailsOn("delete", "servicemeshmemberrolls"),
+		},
+		{
+			name:             "update-member-fails-during-delete",
+			memberRollExists: true,
+			deletion:         true,
+			reactor:          test.ClientFailsOn("update", "servicemeshmembers"),
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			var objects []runtime.Object
+
+			member := newMember()
+			if tc.deletion {
+				member.DeletionTimestamp = &oneMinuteAgo
+			}
+			objects = append(objects, member)
+
+			if tc.memberRollExists {
+				memberRoll := newMemberRoll()
+				if !tc.memberRollCreatedManually {
+					memberRoll.Annotations = map[string]string{
+						common.CreatedByKey: controllerName,
+					}
+				}
+				if tc.deletion {
+					memberRoll.Spec.Members = []string{appNamespace}
+				}
+				objects = append(objects, memberRoll)
+			}
+
+			_, tracker, r := createClientAndReconciler(t, objects...)
+			tracker.AddReactor(tc.reactor)
+
+			assertReconcileFails(r, t)
+		})
+	}
+}
+
+func newMember() *maistra.ServiceMeshMember {
+	return &maistra.ServiceMeshMember{
+		ObjectMeta: meta.ObjectMeta{
+			Name:       memberName,
+			Namespace:  appNamespace,
+			Finalizers: []string{common.FinalizerName},
+		},
+		Spec: maistra.ServiceMeshMemberSpec{
+			ControlPlaneRef: maistra.ServiceMeshControlPlaneRef{
+				Name:      controlPlaneName,
+				Namespace: controlPlaneNamespace,
+			},
+		},
+	}
+}
+
+func newMemberRoll() *maistra.ServiceMeshMemberRoll {
+	return &maistra.ServiceMeshMemberRoll{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      memberRollName,
+			Namespace: controlPlaneNamespace,
+		},
+		Spec: maistra.ServiceMeshMemberRollSpec{
+			Members: []string{},
+		},
+	}
+}
+
+func createClientAndReconciler(t *testing.T, clientObjects ...runtime.Object) (client.Client, *test.EnhancedTracker, *MemberReconciler) {
+	s := scheme.Scheme
+	if err := apis.AddToScheme(s); err != nil { // scheme must be initialized before creating the client below
+		t.Fatalf("Could not add to scheme: %v", err)
+	}
+
+	tracker := clienttesting.NewObjectTracker(scheme.Scheme, scheme.Codecs.UniversalDecoder())
+	enhancedTracker := test.NewEnhancedTracker(tracker)
+	cl := fake.NewFakeClientWithSchemeAndTracker(scheme.Scheme, &enhancedTracker, clientObjects...)
+	fakeEventRecorder := &record.FakeRecorder{}
+
+	r := &MemberReconciler{ResourceManager: common.ResourceManager{Client: cl, PatchFactory: common.NewPatchFactory(cl), Log: log}, scheme: s, eventRecorder: fakeEventRecorder}
+	return cl, &enhancedTracker, r
+}
+
+func assertReconcileSucceeds(r *MemberReconciler, t *testing.T) {
+	res, err := r.Reconcile(request)
+	if err != nil {
+		t.Fatalf("Reconcile failed: %v", err)
+	}
+	if res.Requeue {
+		t.Error("Reconcile requeued the request, but it shouldn't have")
+	}
+
+	// TODO: check if Reconcile mutated the object in the local cache (it should never do that)
```

Comment:
When using the cache directly, you're supposed to clone the object before mutating it, since cached objects are shared by all the controllers running in the same process. Luckily, the controller-runtime client itself already clones the object, so even if we mutate it, we're not doing anything wrong. I've removed the TODO now.

Reply:
Good to know, thanks! :+1: